chain/indexer/src/streamer/streamer.rs

```diff
@@ -84,38 +92,73 @@ async fn fetch_block_by_height(
+/// This function supposed to return the entire `StreamerMessage`.
+/// It fetches the block and all related parts (chunks, outcomes, state changes etc.)
 /// and returns everything together in one struct
-async fn fetch_block_response(
+async fn build_streamer_message(
     client: &Addr<near_client::ViewClientActor>,
     block: views::BlockView,
-    outcomes_to_get: impl Iterator<Item = types::TransactionOrReceiptId>,
-) -> Result<FetchBlockResponse, FailedToFetchData> {
+) -> Result<StreamerMessage, FailedToFetchData> {
     let chunks = fetch_chunks(&client, &block.chunks).await?;
-    let (outcomes, mut outcomes_to_retry) = fetch_outcomes(&client, outcomes_to_get).await;
 
-    for chunk in &chunks {
-        outcomes_to_retry.extend(chunk.transactions.iter().map(|transaction| {
-            types::TransactionOrReceiptId::Transaction {
-                transaction_hash: transaction.hash.clone(),
-                sender_id: transaction.signer_id.to_string(),
-            }
-        }));
+    let mut local_receipts: Vec<views::ReceiptView> = vec![];
+    let mut outcomes = fetch_outcomes(&client, block.header.hash).await?;
+    let mut indexer_chunks: Vec<IndexerChunkView> = vec![];
 
-        outcomes_to_retry.extend(chunk.receipts.iter().map(|receipt| {
-            types::TransactionOrReceiptId::Receipt {
-                receipt_id: receipt.receipt_id,
-                receiver_id: receipt.receiver_id.to_string(),
-            }
-        }));
+    for chunk in chunks {
+        let views::ChunkView { transactions, author, header, receipts } = chunk;
+
+        let indexer_transactions = transactions
+        .into_iter()
+        .map(|transaction| {
+            IndexerTransactionWithOutcome {
+            outcome: outcomes.remove(&transaction.hash).expect("The transaction execution outcome should always present in the same block as the transaction itself"),
+            transaction,
+        }
+        })
+        .collect::<Vec<IndexerTransactionWithOutcome>>();
+
+        local_receipts.extend(
+            indexer_transactions
+                .iter()
+                .filter(|tx| tx.transaction.signer_id == tx.transaction.receiver_id)
+                .map(|tx| views::ReceiptView {
+                    predecessor_id: tx.transaction.signer_id.clone(),
+                    receiver_id: tx.transaction.receiver_id.clone(),
+                    receipt_id: tx
+                        .outcome
+                        .outcome
+                        .receipt_ids
+                        .first()
+                        .expect("The transaction outcome should always include single `receipt_id` in vec")
+                        .clone(),
+                    receipt: views::ReceiptEnumView::Action {
+                        signer_id: tx.transaction.signer_id.clone(),
+                        signer_public_key: tx.transaction.public_key.clone(),
+                        gas_price: block.header.gas_price, // TODO: fill it
```

Comment:
Can't figure out how to use `tx_cost` function here, what `gas_price` to pass to that function. Also I'm dealing with `SignedTransactionView` while function requires `&Transaction`. @evgenykuzyakov can you help here?

Reply:
This seems like a pain.

1. The first thing is to convert `SignedTransactionView` to `SignedTransaction` or to `Transaction`. This can be manually done because you can convert all actions form `ActionView` to `Action` using `try_from`.

2. Use `tx_cost`. Grab `Config` from client, the current protocol version from epoch (@bowenwang1996 ), and `gas_price: block.header.gas_price,` and `sender_is_receiver` is `true` (for local receipt).

3. Finally, you need to get a `receipt_id`. You can grab it from `ExecutionOutcome` from `SuccessReceiptId`

Reply:
which block is this `block`? Is it the block that contains the transaction? If you you need gas_price from previous block as the baseline for calculating receipt gas price.

Reply:
@bowenwang1996 yeah, this is the block containing the transaction. Anyway I need `gas_price` from prev block to pass to `tx_cost` function? Prev block's `gas_price` is not what I should pass here?

Reply:
> Anyway I need gas_price from prev block to pass to tx_cost function? Prev block's gas_price is not what I should pass here?

Yes that is correct. You should use the result from `tx_cost`.